winforms/src/toga_winforms/widgets/base.py

```diff
@@ -48,9 +32,41 @@ def update_scale(self, screen=None):
         hMonitor = windll.user32.MonitorFromRect(screen_rect, 2)
         pScale = wintypes.UINT()
         windll.shcore.GetScaleFactorForMonitor(c_void_p(hMonitor), byref(pScale))
-        Scalable._dpi_scale = pScale.value / 100
-        if not hasattr(Scalable, "original_dpi_scale"):
-            Scalable.original_dpi_scale = Scalable._dpi_scale
+        return pScale.value / 100
+
+    @property
+    def dpi_scale(self):
+        if (
+            hasattr(self, "interface")
+            and (self.interface is not None)
+            and hasattr(self.interface, "window")
+            and (self.interface.window is not None)
+        ):
+            # For Widgets and Stack Trace Dialogs
+            if issubclass(type(self), Widget) or (  # pragma: no branch
+                hasattr(self.interface.window._impl, "current_stack_trace_dialog_impl")
+                and (
+                    self.interface.window._impl.current_stack_trace_dialog_impl == self
+                )
+            ):
+                self._original_dpi_scale = (
+                    self.interface.window._impl._original_dpi_scale
+                )
+                return self.interface.window._impl._dpi_scale
+        # For Containers
+        if hasattr(self, "native_content"):
+            _dpi_scale = self.get_dpi_scale(Screen.FromControl(self.native_content))
+        else:
+            # For Windows and others
+            _dpi_scale = self.get_dpi_scale(Screen.FromControl(self.native))
```

Comment:
Screen.FromControl says it returns "the display that contains the largest portion of the specified control". Can you confirm that this is consistent with how Windows decides to scale the window's title bar? As opposed to, for example, always choosing the display that contains the window's top-left corner.

Reply:
Doing `Screen.FromControl(self.native_content)` on a Container is not valid, because a window may have multiple Containers, e.g. for the SplitContainer, and they should all have the same DPI as the window.

Note that Window and ScrollContainer actually inherit from Container, while OptionContainer and SplitContainer maintain multiple separate Containers.

Reply:
>Can you confirm that this is consistent with how Windows decides to scale the title bar? As opposed to, for example, always scaling it for the display that contains the window's top-left corner.

From my testing, Screen.FromControl indeed returns the screen having the largest portion of the control.

>Also, doing Screen.FromControl(self.native_content) on a Container is not valid, because a window may have multiple Containers, and they should all have the same DPI as the window.

That value is only used when the container widget has no assigned window. Once a window is assigned, they use the DPI of their assigned window.

Reply:
OK, that seems consistent with my testing too, though I've found a few bugs with moving between screens which I'll note below.